File: backend/app/db/models/paper.py
from sqlalchemy import Column, BigInteger, String, Text, Date, DateTime, Boolean, Float, JSON
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from typing import Optional
from app.db.database import Base

class Paper(Base):
    __tablename__ = "papers"

    id = Column(BigInteger, primary_key=True, autoincrement=True)
    external_id = Column(String(100), unique=True, nullable=False, index=True)
    title = Column(Text, nullable=False)
    abstract_en = Column(Text, nullable=False)
    abstract_vi = Column(Text, nullable=True)
    authors = Column("authors", JSON, nullable=True)
    published = Column(Date, nullable=True, index=True)
    source_url = Column(Text, nullable=True)
    pdf_path = Column(String(500), nullable=True)
    source = Column(String(100), nullable=False)
    score = Column("score", Float, nullable=False, default=0, index=True)
    has_audio = Column(Boolean, default=False, nullable=False)
    created_at = Column(DateTime, server_default=func.now(), nullable=False)




    # Relationships
    digest_papers = relationship("DigestPaper", back_populates="paper", cascade="all, delete-orphan")
    audio_abstract = relationship("AudioAbstract", back_populates="paper", uselist=False, cascade="all, delete-orphan")
    chat_sessions = relationship("ChatSession", back_populates="paper", cascade="all, delete-orphan")
    paper_topics = relationship("PaperTopic", back_populates="paper", cascade="all, delete-orphan")

    @property
    def audio_abstract_path(self) -> Optional[str]:
        if self.audio_abstract:
            return self.audio_abstract.file_path
        return None

    @property
    def audio_abstract_url(self) -> Optional[str]:
        if self.audio_abstract and self.audio_abstract.file_path:
            return "/static/" + self.audio_abstract.file_path.lstrip("/")
        return None

    @property
    def audio_duration_seconds(self) -> Optional[float]:
        if self.audio_abstract:
            return self.audio_abstract.duration_seconds
        return None

    @property
    def pdf_url(self) -> Optional[str]:
        if not self.pdf_path:
            return None
        path_str = self.pdf_path.lstrip("/")
        if path_str.startswith("data/"):
            return "/static/" + path_str
        else:
            return "/static/data/" + path_str

File: frontend/src/types/paper.ts
// Types matching backend PaperResponse schema (paper_schema.py)

export interface AudioAbstract {
  id: number;
  file_path: string;
  audio_url: string | null;
  duration_seconds: number | null;
  paper_timestamps: unknown | null;
  created_at: string; // ISO datetime string
}

export interface Paper {
  id: number;
  arxiv_id: string;
  title: string;
  abstract: string;
  summary: string | null;
  /** Backend returns List[str] but may also be null */
  authors: string[] | null;
  published: string | null; // ISO date "YYYY-MM-DD"
  pdf_path: string | null;
  pdf_url: string | null;
  score: number;
  has_audio: boolean;
  created_at: string; // ISO datetime string
  audio_abstract: AudioAbstract | null;
  audio_abstract_path: string | null;
  audio_abstract_url: string | null;
  audio_duration_seconds: number | null;
}

/** GET /papers returns Paper[] directly (no wrapper) */
export type PapersListResponse = Paper[];
